Retrofit haikuports patchset to support haiku as a separated platform from BeOS

thibaultduponchelle opened the pull request:
Hello 😀 

I propose this change to retrofit the portions of haikuports patch related to EUMM. See [haikuports perl patchset](https://github.com/haikuports/haikuports/blob/master/dev-lang/perl/patches/perl-5.32.1.patchset#L186)

Haiku has now some differences versus BeOS (filesystem hierarchy conventions...) that justifies this new "platform variant".

I tested it with perl 5.34.0 but I'm not a haiku expert and the code is not mine but from @olta

@bingos How do you feel about this change ?

Leont commented on lib/ExtUtils/MM_Haiku.pm:
```diff
@@ -0,0 +1,81 @@
+require ExtUtils::MM_Any;
+require ExtUtils::MM_Unix;
+
+our @ISA = qw( ExtUtils::MM_Any ExtUtils::MM_Unix );
```
No. Just no.

I really don't think we should be deriving form `ExtUtils::MM_Any` here. Let's not propagate that headache.

thibaultduponchelle replied:
Fixed

Leont commented on lib/ExtUtils/MM.pm:
```diff
@@ -61,7 +61,8 @@ if( $^O eq 'MSWin32' ) {
 $Is{Cygwin} = $^O eq 'cygwin';
 $Is{NW5}    = $Config{osname} eq 'NetWare';  # intentional
-$Is{BeOS}   = ($^O =~ /beos/i or $^O eq 'haiku');
+$Is{BeOS}   = $^O =~ /beos/i;
```
Does keeping `$Is{BeOS}` even make sense?

thibaultduponchelle replied:
I don't know, I should ask

Leont replied:
I suppose we should keep it for now, and re-evaluate when we look at  #290

Leont commented on lib/ExtUtils/MM_Haiku.pm:
```diff
@@ -0,0 +1,81 @@
+    my $prefixGiven = $self->{PREFIX};
+	$self->SUPER::init_INSTALL_from_PREFIX();
+	if ($prefixGiven) {
+        $self->{SITEPREFIX} = '$(PREFIX)/non-packaged';
```
This doesn't make sense to me. I'm suspecting they're using this the wrong way.

thibaultduponchelle replied:
I have the same hesitation about this line

thibaultduponchelle replied:
I removed this portion of the patch

Leont commented on lib/ExtUtils/MM_Haiku.pm:
```diff
@@ -0,0 +1,81 @@
+sub init_others {
+    my $self = shift;
+
+	$self->SUPER::init_others();
```
The whitespace is inconsistent in various places.

thibaultduponchelle replied:
ouch sorry for that

thibaultduponchelle replied:
fixed

Leont commented on lib/ExtUtils/MM_Haiku.pm:
```diff
@@ -0,0 +1,81 @@
+package ExtUtils::MM_Haiku;
+
+use strict;
+
```
`use warnings;`?

thibaultduponchelle replied:
fixed thx

Leont commented on lib/ExtUtils/MM_Haiku.pm:
```diff
@@ -0,0 +1,81 @@
+
+use ExtUtils::MakeMaker::Config;
+use File::Spec;
+require ExtUtils::MM_Any;
```
These two can just be a `use ExtUtils::MM_Unix;`

thibaultduponchelle replied:
I fixed it, but notes in my update I kept the `require` instead of `use` (is it good?)

Leont replied:
It's not wrong, just a little silly

Leont commented:
This should also adapt `t/MM_BeOS.t`

thibaultduponchelle commented:
I removed a portion of the patch since it should be fixed at perl5 compilation level: if you use -Dprefix... set also siteprefix and you're all good, and if not just use default `./Configure -des` that will set correctly the site prefix to a non-packaged and writable directory :+1:

I changed the `require` to `use`. 

I kept the first function to use vendor on env var, it seems intended to haiku packagers (can be emulated easily for one package with `perl Makefile.PL INSTALLDIRS=vendor`) and I also kept the LD_RUN_PATH patch that seems the most interesting (even if I don't understand it 100%).

Concerning the BeOS support, I'm asking a status to some people from Haiku community to have a first idea, but maybe we can keep this part out of this PR?

thibaultduponchelle commented:
Having continued to work on this topic, I'm not 100% convinced about this change, so I will close it for now, waiting to know what is really needed and what is not.

(Thank you for the code review)
